### ISSUE TITLE
Add feature to mount folders post-provisioning?

### ISSUE THREAD

**Issue by tmotley** [hashicorp/vagrant#936]
I don't think this feature or config setting exists having perused the docs and support.  My Vagrant file mounts shared folders, and assigns them a group and user ownership. However, if the case of a new box, the user and group I assign the mounted folder ownership do not exist until the chef provisioning is done on the box.  I tried to control the order of provisioning and mounting, but the mounting is always attempted first, which obviously breaks if the provisioning has not occurred yet.

For example:

config.vm.define :dpils_slave_1, :primary => false do |cfg|
    cfg.vm.forward_port 8080, 8081
    #cfg.vm.network :hostonly, "1.1.1.11"
    cfg.vm.provision :chef_solo do |chef|
      chef.cookbooks_path = ["chef/cookbooks", "chef/site-cookbooks", "chef/site-cookbooks-tk"]
      chef.roles_path = ["chef/roles-tk"]
      chef.add_role "jboss"
      # chef.log_level = :debug
    end

```
cfg.vm.share_folder 'crx_node_2', '/foo/bar/dpils_repository/dpils_whq_02/crx', 'foo/bar/crx_node_2', :owner => "jboss", :group => "jboss", :create => true
```

  end

Which gives this error on a clean vagrant load:

[dpils_slave_1] -- v-csr-4: /tmp/vagrant-chef-1/chef-solo-4/roles
[dpils_slave_1] -- v-csc-1: /tmp/vagrant-chef-1/chef-solo-1/cookbooks
[dpils_slave_1] -- v-csc-2: /tmp/vagrant-chef-1/chef-solo-2/cookbooks
[dpils_slave_1] -- v-csc-3: /tmp/vagrant-chef-1/chef-solo-3/cookbooks
[dpils_slave_1] -- crx_node_2: /nike/vault/dpils_repository/dpils_whq_02/crx
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

mount -t vboxsf -o uid=`id -u jboss`,gid=`id -g jboss` crx_node_2 /nike/vault/dpils_repository/dpils_whq_02/crx

**Comment by Ghoughpteighbteau** [comment]
I hit this issue as well. Deving a server applet for tomcat, would like to mount the webapps folder so I can do partial compiles from my dev environment, but I can't mount a webapps folder that doesn't exist. Doc's had no answers for me. This is a needed feature. 

+1

**Comment by danomatic** [comment]
+1 for tomcat and other directories that require provisioned users!

**Comment by liaodrake** [comment]
+1

**Comment by fub-frank** [comment]
+1

**Comment by rdubigny** [comment]
+1

**Comment by ryakad** [comment]
+1

**Comment by gaggle** [comment]
+1

**Comment by mturner** [comment]
+1

**Comment by Rican7** [comment]
+1

**Comment by joelesalas** [comment]
+1 on this, becomes a nightmare if you try to use cachefilesd to optimize for performance.

**Comment by clintharris** [comment]
+1

**Comment by delitescere** [comment]
+1

I want to share my host user's ~/.ssh files as the VM vagrant user's ~/.ssh files. The simplest way to do this is to sync the host user's ~/.ssh directory. However, Vagrant needs ~/.ssh to create and manage an authorized_keys file.

A provisioning step I have is to set sshd to use another location for user authorized keys (e.g. /etc/ssh/authorized_keys/%u) and move the vagrant user's authorized_keys file to there (e.g. /etc/ssh/authroized_keys/vagrant). However, the authorized_keys file is destroyed by the folder sync before provisioning has a chance to move it.

I can totally ugly hack this with the following in my Vagrantfile
`config.vm.synced_folder "#{HOST_HOME}/.ssh/", "/home/vagrant/.ssh/" if Pathname.new("./.vagrant/machines/virtualbox/action_provision").exist?` but I do not want.

**Comment by kung-fu-tzu** [comment]
+1

**Comment by ben-rosio** [comment]
Hey @delitescere, just a thought about how you might sidestep the issue in your specific usecase.  I don't know if you've already looked at setting guest machine's ssh to enable [ssh forward_agent](https://developer.github.com/guides/using-ssh-agent-forwarding/).  
You would need an ssh agent locally like Putty's pageant or Mac OS's Keychain (for Leopard and above), and to set [config.ssh.forward_agent](http://docs.vagrantup.com/v2/vagrantfile/ssh_settings.html) to true in your Vagrantfile.

After that, all your local ssh keys will be available on the vagrant box!  You can even put the ssh.forward_agent config in your user Vagrantfile (located in ~/.vagrant.d/Vagrantfile) so it applies to all your vagrant instances.

**Comment by aggrrrh** [comment]
+1

**Comment by dreynolds** [comment]
@mitchellh Is there any chance you might reconsider this? There's been quite a number of +1s

**Comment by Ghoughpteighbteau** [comment]
Even a direct implementation of this feature isn't a requirement. 

The +1's here only indicate a problem, that the owner of a folder might be someone who is only created after the provisioning step has happened. 

If you compare and contrast a Dockerfile, you see why this isn't a problem for them, they can specify any order they please. Maybe that's the real bug here. Users need more control than the Vagrantfile is offering them.

**Comment by ben-rosio** [comment]
Another interesting solution would be to have sync folders be provisioning
steps, this way they can happen in any order necessary.

**Comment by Ghoughpteighbteau** [comment]
@ben-rosio well, that's the hack, right? Adding a post provisiong call in a shell script to mount external folders, and provide an unmounted shared folder at the start so that virtual box knows what to call it.

The problem is it's very non-obvious and strange (and personally I couldn't even get it to work, I ended up just taking advantage of the www-group)

**Comment by Hondilla** [comment]
+1

**Comment by tjbenator** [comment]
+1

**Comment by ktriponis** [comment]
+1

**Comment by clok** [comment]
+1

**Comment by derek-adair** [comment]
@Ghoughpteighbteau -> nail_on_head.gif.  Expecting a Vagrantfile to execute sequentially seems like a far more intuitive UX.  I've not run into this yet, but i'm glad i checked this out b/c i'm sure at some point when i start playing with chef I may.  

Although, i pre-provision my boxes w/ packer so I feel like this issue is largely avoided by that.  It seems like any vagrant provisioning I will ever need will be any experimental packages and will not have any mounting issues.  I'm sure i'm missing some edge cases...

Nice to have, but definitely not a show stopper IMHO.

**Comment by daithi-coombes** [comment]
+1

I would consider this a `bug/blocker` from the standalone pov of `vagrant` and folder syncing. The way I would see it is that 100% of user's created by `package` installation - such as my `jenkins` user - are not able to work with vagrants  `sync_folder` 100% of the time.

But whether its called a bug or enhancement etc - I'd love to see this issue re-opened

**Comment by AngeIII** [comment]
Mount synced folder (not all, but some exact) after provision is what I need. How to workaround this?

**Comment by Ghoughpteighbteau** [comment]
@AngeIII we've being using www-data as that user/group is often pre-existing in many distributions, and has reasonable security settings. 

In some cases this is exactly what we need, in other cases it's a dirty hack.

Mount a third party folder with www-data as the owner, and give your application group access to www-data. 

If that's a blocker for you, well then, this needs to be fixed :\

**Comment by tuupola** [comment]
+1 for this feature too. Currently can not do something like below because user and group `apache` does not exists before provisioning is run.

    config.vm.synced_folder "logs", "/vagrant/logs", owner: "apache", group: "apache"

**Comment by AngeIII** [comment]
My workaround for this:
```
config.vm.provider "virtualbox" do |v|
      v.customize ["setextradata", :id, "VBoxInternal2/SharedFoldersEnableSymlinksCreate/v-root", "1"]
  end
```
And in provision.sh
```
rm -rf /srv/www/web-dev/content/plugins
rm -rf /srv/www/web-master/content/plugins
ln -ds /srv/www/plugins /srv/www/web-dev/content/plugins
ln -ds /srv/www/plugins /srv/www/web-master/content/plugins
```

Works only if vagrant is up with admin rights (Windows).

**Comment by JimDabell** [comment]
+1

**Comment by amirabiri** [comment]
+1

**Comment by topherchristie** [comment]
+1

**Comment by davidski** [comment]
+1

My use case is to have userapp specific data directories on the host available in the guest. The user accounts only exist after provisioning (chef). The hard wired non-existant uid/gid hack doesn't work under vagrant 1.7.2 and vagrant-vmware-workstation 3.2.6 - the mount step tries to do an `id` against the provided uid/gid, which fails.

There's over three years of comments asking for functionality to address this not-uncommon use case.

**Comment by masonmark** [comment]
+1 for @ben-rosio 's :mount_after_provision => true suggestion.

The comments on this issue prove that it is reasonably common to need to mount shared folders **after** provisioning, so that the user that is supposed to own them can be created first.

Furthermore, the workaround to use the UID/GID doesn't work with the VMWare provider, as @davidski points out.

We are using the VMWare provider and Ansible for provisioning, and the only workaround I could find for us was to configure the synced folder in the usual way:

    config.vm.synced_folder "rollerball/", "/rollerball" # (in the Vagrantfile)

That way, it just gets mounted as owned by the `vagrant` user, and then add steps in our provisioning process to unmount the shared volume, and then mount it again it with the correct ownership:

    sudo mount -t vmhgfs -o uid=`id -u rollerball`,gid=`id -g rollerball` .host:/-rollerball /rollerball

Not the end of the world, but not ideal either.

**Comment by DavidHuie** [comment]
+1

**Comment by delitescere** [comment]
I just want to take a moment to point out two things:

1) all of us who would like this feature are empowered to contribute a solution,
2) Mitchell does a great job and vagrant saves us all a load of time, so the ways to overcome this obstacle are kludgey but joyful.

Cheers


----------------------------------
From the phone

**Comment by Ghoughpteighbteau** [comment]
true, but the problem is if you want to fix this issue, you need a slight rearchitecture of how the VagrantFile is run. It's either that, or a really stupid and kludgey hack where you include a directive for post mounting and pre mounting. 

You see the problem here? Implementing this suggestion either makes the API worse, or breaks compatability. 

It should be seriously considered, and frankly, us users know that we can't rightfully make a change like that. That's why we need repository contributors to look at this issue and open a dialogue with us. Only when we have a plan on how this should be implemented can we go and tell people to implement it, right?

**Comment by delitescere** [comment]
@Ghoughpteighbteau 

You had me at "true", you had me at "true".

I don't believe it requires a re-architecture, but it does mean exposing the "provisioned" state more broadly than it is currently. As far as the Vagrantfile syntax, there are indeed at least two approaches:

1) Limit the "post-provisioning" behaviour to folder syncing
2) Provide a block for arbitrary post-provisioning actions

Perhaps 1) directly addresses this issue, whereas 2) gets Vagrant into puppet/chef territory.

**Comment by dwsteele** [comment]
+1.  The workarounds don't work reliably and I end having to run manual steps each time I create a VM.

**Comment by alextyrey** [comment]
Major +1. NFS will not work unless the relevant directory already exists. Very annoying when you want stuff in `/srv/` synced.

**Comment by davesag** [comment]
+1.

**Comment by laupow** [comment]
+1. I ran into this problem on day 1 of using vagrant; it isn't a non-issue.

**Comment by SaimonL** [comment]
+1 Please make ":mount_after_provision => true" possible

**Comment by klingenm** [comment]
+1

**Comment by boomy** [comment]
+1

**Comment by itsmemattchung** [comment]
+1

**Comment by novalis111** [comment]
+1 For the moment I am using a manual approach placing "mount.flag" in the vagrant root and checking if exists in Vagrantfile to en/disable mounting. Post-provision flag or "mount_after_provision" would solve this.

**Comment by SaimonL** [comment]
My work place uses Vagrant with VMware fusion so I can't even use the workaround specified above :(

**Comment by gopass2002** [comment]
+1

**Comment by MartinLyne** [comment]
+1

**Comment by clintbeacock** [comment]
+1

**Comment by gordon-morehouse** [comment]
3 days after I ever touched Vagrant in my life, I hit this when provisioning testing boxes which need to use a shared folder with millions of files and have it be in their 'apache' group.

**Comment by SaimonL** [comment]
In the end the only option I had left was to create a base image box with all the users already created.

**Comment by klingenm** [comment]
Just use the numeric uid/gid for mounting, they shouldn't change that
often. Works nice for me.

**Comment by davidski** [comment]
As discussed at https://github.com/mitchellh/vagrant/issues/936#issuecomment-104826608, uid/gid does not work for all combinations. vagrant-vmware tries to do an id on the uid and guid, which don't exist.

**Comment by notpeter** [comment]
@SaimonL We too created a basebox where it's just vanilla ubuntu + one user account.  Super overkill, but it was the only reasonable workaround which addressed our needs. Given this issue has been open for >3years, an afternoon with Packer to script creation of basebox(es) might be the easiest way to work around this issue.

**Comment by novalis111** [comment]
I am using a "local.yml" file for paths and the UID+GID of the vagrant process during ansible provisioning. After the initial run i have a script (mount.sh) that creates a 'mount.flag' file to indicate that mounting is now possible and updates the /etc/exports for nfs shares + reloads the vagrant box afterwards. See https://gist.github.com/novalis111/e951a3b7d499ccdcbbb1

**Comment by dalcacer** [comment]
+1

**Comment by AngeIII** [comment]
@mitchellh , maybe re-open a ticket as not resolved? Lot of people asking for this feature. Not all workarounds works fine and it should be more clear to make actions before/after/post/cron and etc.

**Comment by rogerwelin** [comment]
+1

**Comment by rogerwelin** [comment]
Indeed this one should be reopened. Although I was able to hardcode the uid and the gid it's extremely hackish and not convienient when doing the provisioning. mount after provision functionality pls!

**Comment by boomy** [comment]
@mitchellh What's the state here? Will this ticket be reopened or not? Many people are asking for this feature.

**Comment by SaimonL** [comment]
I got the best alternative solution. 
Use Chef Solo or Puppet Apply and create a the user and do a NFS mount and set up Autofs so on boot it will auto-mount.
The downside is you will have to setup a NFS server but then again NFS server works for all VMware, VirtualBox, and KVM (vagrant-libvirt) and can be use with other services.

**Comment by slaney** [comment]
@SaimonL I am glad that's working for you, but that breaks the portability of the Vagrantfile.  Unless you now add another machine to it that will act as the NFS server.  I don't think that would be a great alternative for most folks.

**Comment by SaimonL** [comment]
I just realized that Vagrant is all about having everything in one place, simple and easy to set-up and getting started. NFS server takes that away and adds dependencies. So I vote back to  ":mount_after_provision => true" :)

**Comment by Fannon** [comment]
+1

**Comment by wvangeit** [comment]
+1

**Comment by gawbul** [comment]
@notpeter Packer has been our work around for this too.

**Comment by noilly** [comment]
+1

**Comment by raupie** [comment]
+1000   having symfony2 permissions with writing to cache & log folder. apache doesn't exists until after the provisioning is complete.

**Comment by ak5** [comment]
+1

**Comment by flaugher** [comment]
I +1 this.  I've wasted an enormous amount of time looking for a suitable workaround.  One of the main reasons to create a virtual box with Vagrant is so that we can do development in an environment that duplicates our production environment.  If directories where our code resides aren't able to have the same owners and permissions as in our production environment, we're not really duplicating that production environment and are thus missing out on one of the main advantages to using a VM.

**Comment by dreki** [comment]
Here's the workaround I landed on:
```
config.vm.synced_folder /opt/somedir, /opt/somedir, mount_options: ["uid=1010,gid=1010"]
```
To do this you must create the desired user and group with the UID specified (`useradd -u 1010 cooluser` in this case). The folder will be synced, and when the user is actually created, its UID will match and it will have permission on the synced dir.

**Comment by kainjow** [comment]
Adding myself to the list that would like this feature. I am using [vagrant-bindfs](https://github.com/gael-ian/vagrant-bindfs) and need to set the group/owner of the synced folder but need for this to happen after provisioning, which creates the user/group.

**Comment by mikenerone** [comment]
@mitchellh Could Hashicorp just fix the Vagrant provider so that numeric owner and group IDs in the shared_folder config work right (i.e. detect a numeric id and leave out the `id -u/g` bit)? The recent updates to the VirtualBox provider in #7720 are a good template for properly handling all the cases.

**Comment by tareko** [comment]
Anything further with this? I have a slightly different problem that results in the same thing. I want to use the debian/jessie64 vbox, but it doesn't come with the guest additions by default, and thus can't mount vboxfs. As such, I have to install the guest additions before the mount on the first round.

**Comment by nick4fake** [comment]
@tareko The same problem here. Any news?

**Comment by leejo** [comment]
Like @flaugher i've wasted a lot of time trying to work around this so would like to see it supported if possible as it's the last step to getting our vm's in working order from a `vagrant up` without having to run more commands post build.

**Comment by mikenerone** [comment]
@tareko, @nick4fake: For that particular problem, you may find the [vagrant-vbguest plugin](https://github.com/dotless-de/vagrant-vbguest) helpful.

**Comment by erikb85** [comment]
The vbguest plugin should probably be listed in the synced_folders docs. That is basically the behaviour I expect from the core synced_folders module.

**Comment by LoveIsGrief** [comment]
I had to work around this by mounting the folder to a temporary location and then linking to it once the necessary users were created provisioning. I did have to use @jcushman's workaround for mounting though.

I would be great if this were supported

**Comment by jdoss** [comment]
+1 to this issue. 

Currently Fedora 25's rpcbind package no longer starts on boot and if we could provision before synced_folder runs, I could fix it with two shell commands. Instead I have to either wait for the package to be fixed, roll my own vagrant box image with the fix or let the NFS mount timeout, fix the issue and run the provision manually.

It is pretty clear based on the feedback in this issue that there are a handful of use cases for doing a provision before synced_folder.

**Comment by xophere** [comment]
I was wondering if it would be possible for my to precreate my user?  As far as I know my installers don't care if the mysql and apache user already exist.  So if I could get vagrant to precreate them in the way that it is stuffing in virtual box tools before the mount this would be solved for me.  This also create the best case scenario where the install dumps in the default stuff into the directories rather the having something dropped on top of that.

**Comment by xophere** [comment]
So this is not ideal but sort of handles it if I do this with a wrapper.

https://github.com/oriceon/vagrant-virtual-machine-for-web-development/wiki/Repack-box-from-existing-one

But it forces me to hide the vagrant work in a wrapper script for the developer.

**Comment by mikenerone** [comment]
I have a solution to running something post-mounts that's been working for me. I'm using the "always" provisioning feature (I'm not sure when that was added):
```
Vagrant.configure('2') do |config|
    ...
    config.vm.provision :shell, path: './my_real_provisioning_script.sh'
    config.vm.provision :shell, run: 'always', path: './my_after_mounts_every_boot_script.sh'
    ...
```
I hope this helps someone out.

**Comment by cecilemuller** [comment]
Personally I was using [this trick](http://stackoverflow.com/questions/24855635/check-if-vagrant-provisioning-has-been-done) to detect first run:

````ruby
if File.exist?(".vagrant/machines/YOUR_BOX_ID/virtualbox/action_provision")
	#
	# Already provisioned, it's now safe to use "config.vm.synced_folder" statements
	# and additional scripts that require the users exist.
	#
else
	#
	# Not yet provisioned, only run the script that creates the required users
	#
end
````

The one downside is that you have to do this once:

	vagrant up
	vagrant reload

Also, if you want the additional scripts to run only once (like provision scripts are meant to), use `vagrant reload --provision` instead of just `vagrant reload`. Otherwise, the scripts need to use `run:always` mentioned in the previous comment.

**Comment by quixand** [comment]
This is rubbish, please fix.

**Comment by snfnwgi** [comment]
Vagrant attempted to execute the capability 'mount_nfs_folder'
on the detect guest OS 'windows', but the guest doesn't
support that capability. This capability is required for your
configuration of Vagrant. Please either reconfigure Vagrant to
avoid this capability or fix the issue by creating the capability.

**Comment by xophere** [comment]
Nope guest is was Linux.

**Comment by snfnwgi** [comment]
I installed windows on the windows, I would like to use NFS way to connect WINDOWS

**Comment by xophere** [comment]
Oh my bad I thought you were replying.  This is really a totally different
feature then my request.

**Comment by snfnwgi** [comment]
@xophere 
boot error! Sorry, my English is not good！
==> default: Forwarding ports...
    default: 5985 (guest) => 55985 (host) (adapter 1)
    default: 5986 (guest) => 55986 (host) (adapter 1)
==> default: Running 'pre-boot' VM customizations...
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: WinRM address: 127.0.0.1:55985
    default: WinRM username: IEUser
    default: WinRM execution_time_limit: PT2H
    default: WinRM transport: plaintext
==> default: Machine booted and ready!
==> default: Checking for guest additions in VM...
==> default: Configuring and enabling network interfaces...
==> default: Exporting NFS shared folders...
==> default: Preparing to edit /etc/exports. Administrator privileges will be required...
==> default: Mounting NFS shared folders...
Vagrant attempted to execute the capability 'mount_nfs_folder'
on the detect guest OS 'windows', but the guest doesn't
support that capability. This capability is required for your
configuration of Vagrant. Please either reconfigure Vagrant to
avoid this capability or fix the issue by creating the capability.

**Comment by xophere** [comment]
You have to load special software for Windows to talk NFS and no one really
does this any more.   You would be better off using samba for windows.  Why
inject NFS?  this is the kind of thing Vagrant isn't good at and probably
shouldn't be.

**Comment by snfnwgi** [comment]
@xophere I use the MAC OS, I want to work in the MAC installed windows7, the default way to visit very slow, so I would like to use nfs

**Comment by xophere** [comment]
I would try a smb share from the Mac.  But again if that is what you want
you should open a new request.

**Comment by snfnwgi** [comment]
@xophere Thank you, there are ways to support nfs? I have tried to use SMB and also reported some errors

**Comment by snfnwgi** [comment]
@xophere SMB shared folders are only available when Vagrant is running
on Windows. The guest machine can be running non-Windows. Please use
another synced folder type.

**Comment by snfnwgi** [comment]
@xophere smb error;
➜  modernie-winrm git:(master) ✗ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
SMB shared folders are only available when Vagrant is running
on Windows. The guest machine can be running non-Windows. Please use
another synced folder type.

**Comment by xophere** [comment]
_START A NEW REQUEST!_. PLEASE

**Comment by snfnwgi** [comment]
@xophere  here: https://github.com/hashicorp/vagrant/issues/9008  thank you!

**Comment by snfnwgi** [comment]
@xophere You give me a message, I want to rest, very late, tomorrow to work, thanks, cheers

**Comment by egnikitin** [comment]
https://github.com/gael-ian/vagrant-bindfs hels mount after provision

**Comment by najamelan** [comment]
Ok,

Im going to find a workaround for this. So far I have found some tools that can help:

- vagrant reload plugin: allows to reload between provisioning steps  (although there seem to be alot of issues with this plugin). as an alternative, maybe run halt in the guest and do `vagrant up && vagrant provision`
- vagrant trigger: run code before or after vagrant commands
- ruby code in Vagrantfile can detect environment variables, or run shell commands with the  vagrant-host-shell plugin, or test for the existance of files on the host etc...

Idea:

- have the synced folder in conditional code
- after vagrant up, set an indicator on the host (touch a file or an env variable)
- after first provision step, reload
- now detect first provision has been done
- before vagrant destroy, remove the indicator file

**Comment by d4rkd0s** [comment]
> If you omit the guest path, it won't auto-mount. Also, the @dominis workaround is quite good. I think this is rare enough that this is satisfactory for now since this would require significant change. Sorry!
@mitchellh
Does @dominis issue cover VMWare? I see `vbox` in there, and I get wondering. What would this "solution" look like for VMWare?

**Comment by d4rkd0s** [comment]
I found the GID UID works. If anyone is interested in that solution.

Just make sure you create your user's in your provisioning with the optional `uid` and `gid` for the group. I use Ansible so for example it's this for my `wildfly` user and group:
```
- name: Create wilfdly group
  group:
    name: wildfly
    gid: 1001
    state: present

- name: Create wildfly user
  user:
    name: wildfly
    uid: 1001
    comment: Created by ansible
    group: wildfly
    home: /home/wildfly
```

And then in my `Vagrantfile`:
```
WILDFLY_UID = 1001
WILDFLY_GID = 1001
api.vm.synced_folder "C:/work/war", "/opt/api/deployments", type: "rsync", rsync__auto: true, create: true, owner: WILDFLY_UID, group: WILDFLY_GID
```

**Comment by cbj4074** [comment]
7.5 years later, I'm wondering if there is a new/better "solution" to this problem than those discussed previously.

I've read every comment and my use-case is most similar to the scenario @jdoss describes in https://github.com/hashicorp/vagrant/issues/936#issuecomment-276864313 and @najamelan describes in https://github.com/hashicorp/vagrant/issues/936#issuecomment-454798627 .

In short, the problem I face is not that the user as whom the share(s) should be mounted does not yet exist; rather, the problem is that the networking configuration required to facilitate the sharing is not in place by the time Vagrant tries to auto-mount the shares.

Specifically, I need to assign a static IP address in the VM, and Vagrant does not support this capability the Hyper-V provider (see both https://www.vagrantup.com/docs/hyperv/limitations.html and https://github.com/hashicorp/vagrant/issues/8384 ).

In terms of possible workarounds/solutions to this problem, the "manual" mount command that @dominis describes in https://github.com/hashicorp/vagrant/issues/936#issuecomment-7179034 would be different for Hyper-V machines (i.e., it would not use `-t vboxfs`); I mention this only to underscore the fact that this particular "solution" is provider-specific and therefore less than ideal.

The approach that @cecilemuller describes in https://github.com/hashicorp/vagrant/issues/936#issuecomment-288063253 is promising, but it has two significant drawbacks: a) it's box-name specific, and b) it requires `vagrant up && vagrant reload` to be fully-provisioned.

In essence, I'd really love to be able to defer shared folder mounting until after Vagrant has run an arbitrary provisioner script in which I'm able to do whatever is needed within the guest OS to configure the static IP address.

Ultimately, I'm curious if anybody has found a better way to do this recently... thanks in advance!
